File: app/models/page.rb
require 'journey_questionnaire'

class Page < ActiveRecord::Base
  belongs_to :questionnaire
  has_many :questions, -> { order(:position).includes(:page, :question_options, :special_field_association) }, :dependent => :destroy
  has_many :fields, -> { where(type: Question.field_types.map(&:name)).order(:position) }, :class_name => 'Question'
  has_many :decorators, -> { order(:position).where(type: Question.decorator_types.map(&:name)) }, :class_name => 'Question'
    
  before_create :set_untitled
  before_create :set_position
    
  def number
    questionnaire.pages.index(self) + 1
  end
  
  private
  def set_untitled
    if self.title.blank?
      self.title = "Untitled page"
    end
  end
  
  def set_position
    return if position
    self.position = (questionnaire.pages.maximum(:position) || 0) + 1
  end
end

File: journey-ember/app/models/page.js
import DS from 'ember-data';
import Ember from 'ember';
var Page;

Page = DS.Model.extend({
  questionnaire: DS.belongsTo('questionnaire'),
  questions: DS.hasMany('question', {
    async: true
  }),
  title: DS.attr('string'),
  position: DS.attr('number'),
  questionsSorting: ['position'],
  questionsSorted: Ember.computed.sort('questions', 'questionsSorting'),
  number: (function() {
    var pages;
    pages = this.get('questionnaire.pagesSorted');
    return pages && (pages.indexOf(this) + 1);
  }).property('questionnaire.pagesSorted'),
  isFirstPage: Ember.computed.equal('number', 1),
  isLastPage: (function() {
    return this.get('number') === this.get('questionnaire.pages.length');
  }).property('questionnaire.pages.length', 'number')
});

export default Page;
